shaders/src/light_indirect.fs

```diff
@@ -332,6 +332,114 @@ void evaluateSubsurfaceIBL(const PixelParams pixel, const vec3 diffuseIrradiance
+    const vec3 n0, vec3 E, vec3 Fd, vec3 Fr,
+    inout vec3 color) {
+
+    vec3 r = -shading_view;
+    float eta0 = pixel.etaIR;
+
+    /* compute first interface refraction */
+#if REFRACTION_TYPE == REFRACTION_TYPE_SOLID
+    // for a sphere, with light at infinity (i.e. cubemap)
+    r = refract(r, n0, eta0);
+    float N0oR = dot(n0, r);
+#elif REFRACTION_TYPE == REFRACTION_TYPE_THIN
+    // for thin layers, the direction of the refracted ray is the same as the incoming ray
+#else
+#error "invalid REFRACTION_TYPE"
+#endif
+
+    /* compute distance traveled d */
+    float d = 0.0;
+#if defined(MATERIAL_HAS_ABSORPTION) || (REFRACTION_MODE == REFRACTION_MODE_SCREEN_SPACE)
+#if REFRACTION_TYPE == REFRACTION_TYPE_SOLID
+#if defined(MATERIAL_HAS_THICKNESS)
+    d = pixel.thickness * -N0oR;
+#endif
+#elif REFRACTION_TYPE == REFRACTION_TYPE_THIN
+#if defined(MATERIAL_HAS_MICRO_THICKNESS)
+    // note: we need the refracted ray to calculate the distance traveled
+    // we could use shading_NoV, but we would lose the dependency on ior.
+    float N0oR = dot(n0, refract(r, n0, eta0));
+    d = pixel.uThickness / max(-N0oR, 0.01);
+#endif
+#endif
+#endif
+
+    /* compute transmission T */
+#if defined(MATERIAL_HAS_ABSORPTION)
+#if defined(MATERIAL_HAS_THICKNESS) || defined(MATERIAL_HAS_MICRO_THICKNESS)
+    vec3 T = min(vec3(1.0), exp(-pixel.absorption * d));
+#else
+    vec3 T = 1.0 - pixel.absorption;
+#endif
+#endif
+
+    // Roughness remaping for thin layers, see Burley 2012, "Physically-Based Shading at Disney"
+    //
+    // We apply this remapping to solid objects too because we are handling both interfaces
+    // in one go for solids as well.
+    // Note: this remapping is valid only for realistic IORs.
+    // Note: it's unclear whether we should use perceptualRoughness or roughness here, using
+    // roughness seems to produce worse results.
+    float perceptualRoughness = saturate((0.65 * pixel.etaRI - 0.35) * pixel.perceptualRoughnessUnclamped);
+
+    /* sample the cubemap or screen-space */
+#if REFRACTION_MODE == REFRACTION_MODE_CUBEMAP
+#if REFRACTION_TYPE == REFRACTION_TYPE_SOLID
+    // compute the direction of the output ray (needed for accessing the cubemap)
+    vec3 n1 = normalize(N0oR * r - n0 * 0.5);
+    float eta1 = pixel.etaRI;
+    r = refract(r, n1, eta1);
+#elif REFRACTION_TYPE == REFRACTION_TYPE_THIN
+    // For thin surfaces, the light will bounce off at the second interface in the direction of
+    // the reflection, effectively adding to the specular, but this process will repeat itself.
+    // Each time the ray exits the surface on the front side after the first bounce,
+    // it's multiplied by E^2, and we get: E + E(1-E)^2 + E^3(1-E)^2 + ...
+    // This infinite serie converges and is easy to simplify.
+    // Note: we calculate these bounces only on a single component,
+    // since it's a fairly subtle effect.
+    E *= 1.0 + pixel.transmission * (1.0 - E.g) / (1.0 + E.g);
```

Comment:
Is the difference really worth it?

Reply:
Yeah, I think so, it's very visible at about 45 degrees angle -- in fact, it's not grazing that benefits the most, nor straight; but rather all "medium" angles. For a window, it makes it visibly darker. Also it makes us match mitsuba almost perfectly. given how cheap (and let's face it, elegant) it is, I think it's worth it.

Reply:
Sounds great then!